routing.go

```diff
@@ -669,7 +655,7 @@ func (dht *IpfsDHT) FindPeer(ctx context.Context, id peer.ID) (_ peer.AddrInfo,
 		return pi, nil
```

Comment:
Overall I agree that the API should not guarantee connectivity.

Are you referring to the return statement or to the stop function? For the return statement we could be a little more clever and separately track if any peer addresses are discovered for the peer during the target query, but this falls apart if we separately find the peer via another service while FindPeer is going on.

However, what's the abort condition? If the abort condition is just hearing about the peer then we could end up not being able to connect to the peer because we got a single stale address for them and then gave up.

Once we have signed peer records in place FindPeer should no longer require connectedness (although we could have it as an secondary abort condition). See https://github.com/libp2p/go-libp2p/issues/784 for a little more background on FindPeer thoughts.

Reply:
**1. The return statement:**

```
// TODO: Consider unlucky disconnect timing and potentially utilizing network.CanConnect or something similar
	if dht.host.Network().Connectedness(id) == network.Connected {
		return dht.peerstore.PeerInfo(id), nil
	}

	return peer.AddrInfo{}, routing.ErrNotFound
```

If we have a address for that peer in the peerstore, irrespective of whether we have a connection or not, we should return it. Even if we separately find the peer via another service while FindPeer is going on(low probability), there is no harm in doing this. I feel like mandating connectivity is too restrictive here.

**2. Abort condition**

> If the abort condition is just hearing about the peer then we could end up not being able to connect to the peer because we got a single stale address for them and then gave up.

Sure. Signed peer records will solve this. Are you saying you want to wait till it goes live before you remove this? If yes, that's fine by me.

Reply:
@aarshkshah1992 I modified this a bit lmk if it seems reasonable to you.

1. I ideally want to return routing.ErrNotFound whenever the DHT didn't cause you to find the records. Most people are probably doing FindPeer and then host.Connect, or peerstore.PeerInfo anyhow instead of actually using the records returned here. This is somewhat impractical currently so instead I've thrown in a few conditions that indicate you may have received valid peerinfo (we can only verify via connecting until signed peer records lands) from the DHT query. lmk if you have issues/concerns with my proposal.

2. I think we'll have to discuss this once the signed peer records land. I suspect what we'll end up with is similar to the proposals in https://github.com/libp2p/go-libp2p/issues/784. Have a FindPeerAsync function that returns the records as we get them and perhaps abort early if we're connected since then we know the peerstore definitely has good addresses.